File: apple/HabitKit/Sources/HabitCore/Models/Schedule.swift
/// Wann ein Habit fällig ist.
public enum Schedule: Hashable, Sendable {
    /// Jeden Tag.
    case daily
    /// An festen Wochentagen.
    case weekdays(Set<Weekday>)
    /// n-mal pro Woche, der Tag ist egal. Wird wochenweise ausgewertet.
    case timesPerWeek(Int)
    /// Alle n Tage, gerechnet ab `anchor`.
    case everyNDays(n: Int, anchor: CalendarDate)

    /// Ob ein einzelner Tag verpflichtend ist.
    ///
    /// Bei `timesPerWeek` ist er das nicht: der Habit darf an jedem Tag erledigt
    /// werden, deshalb ist ein Tag ohne Eintrag dort kein verpasster Tag, sondern
    /// gar kein geplanter. Bewertet wird die Woche, nicht der Tag.
    public var requiresSpecificDays: Bool {
        if case .timesPerWeek = self { false } else { true }
    }
}

extension Schedule: Codable {
    private enum CodingKeys: String, CodingKey { case kind, days, n, anchor }
    private enum Kind: String, Codable { case daily, weekdays, timesPerWeek, everyNDays }

    public init(from decoder: any Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        switch try c.decode(Kind.self, forKey: .kind) {
        case .daily:
            self = .daily
        case .weekdays:
            self = .weekdays(Set(try c.decode([Weekday].self, forKey: .days)))
        case .timesPerWeek:
            self = .timesPerWeek(try c.decode(Int.self, forKey: .n))
        case .everyNDays:
            self = .everyNDays(n: try c.decode(Int.self, forKey: .n),
                               anchor: try c.decode(CalendarDate.self, forKey: .anchor))
        }
    }

    public func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .daily:
            try c.encode(Kind.daily, forKey: .kind)
        case .weekdays(let days):
            try c.encode(Kind.weekdays, forKey: .kind)
            try c.encode(days.sorted(), forKey: .days)
        case .timesPerWeek(let n):
            try c.encode(Kind.timesPerWeek, forKey: .kind)
            try c.encode(n, forKey: .n)
        case .everyNDays(let n, let anchor):
            try c.encode(Kind.everyNDays, forKey: .kind)
            try c.encode(n, forKey: .n)
            try c.encode(anchor, forKey: .anchor)
        }
    }
}

/// Wie ein Zielwert zu lesen ist.
public enum Comparison: String, Codable, Sendable, Hashable {
    /// „mindestens" — 2 Liter Wasser trinken.
    case atLeast
    /// „höchstens" — maximal 30 Minuten Social Media.
    case atMost
}

public struct Target: Codable, Hashable, Sendable {
    public var value: Double
    public var unit: String
    public var comparison: Comparison

    public init(value: Double, unit: String, comparison: Comparison = .atLeast) {
        self.value = value
        self.unit = unit
        self.comparison = comparison
    }
}

/// Zeitplan und Ziel gelten ab einem Datum.
///
/// Vergangene Tage werden mit der damals gültigen Regel bewertet. Ohne diese
/// Versionierung würde eine Zielerhöhung von 2 L auf 3 L rückwirkend alle
/// erfüllten Tage als verfehlt erscheinen lassen.
public struct HabitRule: Codable, Hashable, Sendable {
    public var effectiveFrom: CalendarDate
    public var schedule: Schedule
    public var target: Target?

    public init(effectiveFrom: CalendarDate, schedule: Schedule, target: Target? = nil) {
        self.effectiveFrom = effectiveFrom
        self.schedule = schedule
        self.target = target
    }
}

File: server/src/domain/schedule.ts
/// Portiert aus `apple/HabitKit/Sources/HabitCore/Models/Schedule.swift`.
///
/// Swifts `enum` mit assoziierten Werten wird zur unterschiedenen Vereinigung
/// über `kind` — und zwar genau in der Form, die Swifts `Codable` schon
/// schreibt. Damit ist der Typ zugleich das Übertragungsformat, ohne Umweg.

import type { CalendarDate, Weekday } from "./calendar.ts";

/// Wann ein Habit fällig ist.
export type Schedule =
  /// Jeden Tag.
  | { kind: "daily" }
  /// An festen Wochentagen.
  | { kind: "weekdays"; days: Weekday[] }
  /// n-mal pro Woche, der Tag ist egal. Wird wochenweise ausgewertet.
  | { kind: "timesPerWeek"; n: number }
  /// Alle n Tage, gerechnet ab `anchor`.
  | { kind: "everyNDays"; n: number; anchor: CalendarDate };

/// Ob ein einzelner Tag verpflichtend ist.
///
/// Bei `timesPerWeek` ist er das nicht: der Habit darf an jedem Tag erledigt
/// werden, deshalb ist ein Tag ohne Eintrag dort kein verpasster Tag, sondern
/// gar kein geplanter. Bewertet wird die Woche, nicht der Tag.
export function requiresSpecificDays(schedule: Schedule): boolean {
  return schedule.kind !== "timesPerWeek";
}

/// Wie viele Tage einer Woche erledigt sein müssen. Nur bei `timesPerWeek`
/// von Null verschieden.
export function weeklyTarget(schedule: Schedule): number | null {
  return schedule.kind === "timesPerWeek" ? schedule.n : null;
}

/// Wie ein Zielwert zu lesen ist.
export type Comparison =
  /// „mindestens" — 2 Liter Wasser trinken.
  | "atLeast"
  /// „höchstens" — maximal 30 Minuten Social Media.
  | "atMost";

export type Target = {
  value: number;
  unit: string;
  comparison: Comparison;
};

/// Zeitplan und Ziel gelten ab einem Datum.
///
/// Vergangene Tage werden mit der damals gültigen Regel bewertet. Ohne diese
/// Versionierung würde eine Zielerhöhung von 2 L auf 3 L rückwirkend alle
/// erfüllten Tage als verfehlt erscheinen lassen.
export type HabitRule = {
  effectiveFrom: CalendarDate;
  schedule: Schedule;
  target?: Target | null;
};
